File: ios/PlaytomicFoundation/observable/Observable.swift
//
//  Observable.swift
//  Anemone SDK
//
//  Created by Angel Luis Garcia on 19/03/2019.
//  Copyright © 2019 Syltek Solutions S.L. All rights reserved.
//

import Foundation

open class Observable<T> {
    public typealias Observer = (T?) -> Void

    private var observers: [Int: (Observer, DispatchQueue?)] = [:] {
        didSet {
            hasObservers = !observers.isEmpty
        }
    }
    open var hasObservers: Bool = false
    private var uniqueID = (0...).makeIterator()

    fileprivate var _value: T? {
        didSet {
            synchronized(self) {
                let value = _value
                observers.values.forEach { observer, dispatchQueue in
                    if let dispatchQueue = dispatchQueue {
                        dispatchQueue.async {
                            observer(value)
                        }
                    } else {
                        observer(value)
                    }
                }
            }
        }
    }

    open var value: T? {
        _value
    }

    fileprivate init() { }

    open func observe(_ queue: DispatchQueue? = DispatchQueue.main, _ observer: @escaping Observer) -> Disposable {
        synchronized(self) {
            // swiftlint:disable force_unwrapping
            let id = uniqueID.next()!
            // swiftlint:enable force_unwrapping

            observers[id] = (observer, queue)

            let disposable = Disposable { [weak self] in
                self?.observers[id] = nil
            }

            if let value = self.value {
                observer(value)
            }
            return disposable
        }
    }

    open func removeAllObservers() {
        synchronized(self) {
            observers.removeAll()
        }
    }
}

public extension Observable {
    func singleValue() -> Promise<T> {
        Promise<T> { fulfill, _ in
            if let value = self.value {
                fulfill(value)
            } else {
                let disposal = Disposal()
                self.observe { value in
                    if let value = value {
                        disposal.dispose()
                        fulfill(value)
                    }
                }.add(to: disposal)
            }
        }
    }

    func whenValue(condition: @escaping (T) -> Bool) -> Promise<T> {
        Promise<T> { fulfill, _ in
            if let value = self.value, condition(value) {
                fulfill(value)
            } else {
                let disposal = Disposal()
                self.observe { value in
                    if let value = value, condition(value) {
                        disposal.dispose()
                        fulfill(value)
                    }
                }.add(to: disposal)
            }
        }
    }

    func observeNext(_ queue: DispatchQueue? = DispatchQueue.main, _ observer: @escaping Observer) -> Disposable {
        if value == nil {
            return observe(queue, observer)
        } else {
            // Ignore first event since it is a previous value
            var isFirstEvent = true
            return observe(queue) { value in
                if isFirstEvent {
                    isFirstEvent = false
                } else {
                    observer(value)
                }
            }
        }
    }

    func observeChange(_ queue: DispatchQueue? = DispatchQueue.main, observer: @escaping ((T?, T?) -> Void)) -> Disposable {
        var previousValue = value
        return observeNext(queue) { newValue in
            observer(previousValue, newValue)
            previousValue = newValue
        }
    }

    func map<R>(_ map: @escaping (T) -> R) -> Observable<R> {
        MappedObservable(original: self, transform: map)
    }
}

open class MutableObservable<T>: Observable<T> {
    override public init() { }
    public init(value: T?) {
        super.init()
        _value = value
    }

    override open var value: T? {
        get {
            _value
        }
        set {
            _value = newValue
        }
    }
}

public extension MutableObservable where T: Equatable {
    var distinctValue: T? {
        get {
            _value
        }
        set {
            if value != newValue {
                value = newValue
            }
        }
    }
}

class MappedObservable<T, R>: Observable<R> {
    let original: Observable<T>
    let transform: (T) -> R
    init(original: Observable<T>, transform: @escaping (T) -> R) {
        self.original = original
        self.transform = transform
    }

    override var hasObservers: Bool {
        get { original.hasObservers }
        set { original.hasObservers = newValue }
    }

    override var value: R? {
        original.value.map(transform)
    }

    override func observe(_ queue: DispatchQueue? = DispatchQueue.main, _ observer: @escaping Observable<R>.Observer) -> Disposable {
        original.observe { [transform] in
            if let value = $0 {
                observer(transform(value))
            }
        }
    }

    override func removeAllObservers() {
        original.removeAllObservers()
    }

}

File: android/app/src/main/java/com/playtomic/test/foundation/observable/Observable.kt
package com.playtomic.foundation.observable

import android.annotation.SuppressLint
import android.os.Looper
import androidx.annotation.AnyThread
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.map
import com.playtomic.foundation.promise.Promise

typealias Observable<T> = LiveData<T>
typealias MutableObservable<T> = ThreadMutableLiveData<T>

fun <T> Observable<T>.observe(lambda: ((T?) -> Unit)): Disposable {
    val observer = Observer<T>(lambda)
    this.observeForever(observer)
    return Disposable(dispose = { this.removeObserver(observer) })
}

fun <T> Observable<T>.singleValue(): Promise<T> =
    Promise<T> { fulfill, _ ->
        val currentValue = value
        if (currentValue != null) {
            fulfill(currentValue)
        } else {
            val disposal = Disposal()
            this.observe {
                if (it != null) {
                    disposal.dispose()
                    fulfill(it)
                }
            }.add(to = disposal)
        }
    }

fun <T> Observable<T>.whenValue(condition: (T) -> Boolean): Promise<T> =
    Promise<T> { fulfill, _ ->
        val currentValue = value
        if (currentValue != null && condition(currentValue)) {
            fulfill(currentValue)
        } else {
            val disposal = Disposal()
            this.observe {
                if (it != null && condition(it)) {
                    disposal.dispose()
                    fulfill(it)
                }
            }.add(to = disposal)
        }
    }

fun <T> Observable<T>.observeNext(observer: ((T?) -> Unit)): Disposable {
    return if (this.value == null) {
        observe(observer)
    } else {
        // Ignore first event since it is a previous value
        var isFirstEvent = true
        observe { value ->
            if (isFirstEvent) {
                isFirstEvent = false
            } else {
                observer(value)
            }
        }
    }
}

fun <T, R> Observable<T>.map(map: (T?) -> R): Observable<R> = this.map(map)

fun <T> Observable<T>.observeChange(observer: ((T?, T?) -> Unit)): Disposable {
    var previousValue = this.value
    return observeNext { newValue ->
        observer(previousValue, newValue)
        previousValue = newValue
    }
}

open class ThreadMutableLiveData<T> : MutableLiveData<T> {
    constructor() : super()
    constructor(value: T?) : super(value)

    var distinctValue: T?
        get() = value
        set(newValue) {
            if (value != newValue) {
                value = newValue
            }
        }

    @AnyThread
    @SuppressLint("WrongThread", "NullSafeMutableLiveData")
    override fun setValue(value: T?) {
        if (Looper.getMainLooper().thread == Thread.currentThread())
            super.setValue(value)
        else
            super.postValue(value)
    }
}
